WIP: editorial pass to rearrange/layer spec

kimdhamilton opened the pull request:
[See preview](https://centrehq.github.io/presentation-exchange/build/)

Reflects discussion of Jan 20 Discussion. 

**Notes**
1. `limit_disclosure` is now entirely moved to a Feature
2. Rolled logically-similar properties more aggressively into single Features (e.g., Advanced Constraints)
3. Added language to address MUST/MAY topic, e.g., "When using this Feature, the constraints object MAY contain a limit_disclosure property. If present, its value MUST be one of the following strings"
4. I think there still may be some lingering weirdness in Submission Requirements, i.e., if a Feature introduces extensions to multiple objects
5. Tried to add special formatting to Input Evaluation Extension handling. It's awkward and I probably need to discuss with @csuwildcat 

**Issues**
1. Other similar restructuring improvements: consider presentation submission special case (for 1) to avoid need for `descriptor_map`?  Consider doing this in separate PR
2. Need language around processing entity handling when Features not supported, reflecting the "naughty behavior" discussion. i feel like I can't capture this as well as @csuwildcat 
3. TODO: I need to ensure David's examples are captured
4. TODO I manually checked in index.html in my branch; will remove that when merging
5. TODO: Update authors/editors. Add David Chadwick and get updated list in general.

Feel free to refer to Note # or Issue # when commenting. E.g., "Note 2: I disagree"


Fixes #283; Overrides #275

bumblefudge commented:
NB @David-Chadwick

bumblefudge commented:
I've been trying to understand Issue # 2, and i am a little stumped.  Part of me thinks we should spill a paragraph's worth of free digital ink after the second graf of `## Input Evaluation` (L 1191 in the PR), explaining that the choices taken by processors in Input Evaluation may be judged or tracked by Verifiers, and to consider optional features subject to reputation-scoring.  Was this what Daniel meant?  Would a mentioned in the definition of Features in the `## Terminology` section help, since they are free from conformance-stick but still subject to the reputation-carrot?

bumblefudge commented:
Sidenote, are these 5 issues being opened as tracking issues if not addressed by this merge? Not sure if that was implied or not but asking for the sake of clear process.

kimdhamilton commented:
> Sidenote, are these 5 issues being opened as tracking issues if not addressed by this merge? Not sure if that was implied or not but asking for the sake of clear process.

It is a mix of "notes to self" that I will address before merging (e.g. index.html hack), issues we may quickly decide on as a group (and I can address with this change", and (default) issues I will create as separate tracking issues if not address.

kimdhamilton commented:
> I've been trying to understand Issue # 2, and i am a little stumped. Part of me thinks we should spill a paragraph's worth of free digital ink after the second graf of `## Input Evaluation` (L 1191 in the PR), explaining that the choices taken by processors in Input Evaluation may be judged or tracked by Verifiers, and to consider optional features subject to reputation-scoring. Was this what Daniel meant? Would a mentioned in the definition of Features in the `## Terminology` section help, since they are free from conformance-stick but still subject to the reputation-carrot?

Definitely that's the right line of thinking -- this is just an area I hadn't already addressed in the PR but something I want tracked. So worst case, if I don't have time to get to it in this already massive change, we can create a tracking issue

JaceHensley commented on spec/spec.md:
```diff
@@ -836,7 +867,316 @@ processing-related rules above:
+  have high confidence they are also supported by the
+  [[ref:Holder]], before indicating predicate responses are
+  required.:::
+- `preferred` - This indicates that the returned value
```
#supernit maybe the order of `"preferred"` and `"required"` should be the same for properties that follow that pattern?

bumblefudge replied:
that's not a supernit that's a valid question, i'd say

kimdhamilton replied:
Yeah, i'm putting this on hold for now

JaceHensley reviewed (commented):
Issue #1: Can you explain this one more for me? Are you saying there are cases where a `descriptor_map` isn't needed? The only case I can think of that could omit that would be when requesting a single credential but even then I think you'd need to know where to look for the credential in the envelop

JaceHensley commented:
Note # 1: Based on discussion on Jan 27th `limit_disclosure` should be moved into the base profile with additional language around `"required"` ("If you don't support "required" don't return anything" / "Wallets must understand how to at least know when "limit_disclosure" is "required"")

kimdhamilton commented:
Thanks for the feedback! Here are the updates:

- Addressed feedback from Jan 27
- Renamed the now-predicate section to "Predicate Feature"

[See preview](https://centrehq.github.io/presentation-exchange/build/)

I agree with the discussion point that we should land this and follow up with subsequent PRs.

Things that are a prereq for check-in (for my tracking):
- ensure index.html is not in merge

Issues I will create:
- Need a better description of feature behavior, i.e., "light up" if features understood, backwards compatible
- Need language around processing entity handling when Features not supported, reflecting the "naughty behavior" discussion. @csuwildcat 
- (Unless we're ok with it as is) Fix formatting in "Input Evaluation" section for Features.
- Update authors/editors

Other group TODOs:
- review David's original PR and determine if there's anything missing from my check-in and create issues

nklomp commented on spec/spec.md:
```diff
@@ -836,7 +877,298 @@ processing-related rules above:
+      submit a response that has been _self-attested_, i.e., the
+      [[ref:Claim]] used in the presentation was 'issued' by the
+      [[Ref:Subject]] of the [[ref:Claim]].
+    - `preferred` - This indicates that it is ****RECOMMENDED**** that the
```
What is the use case for a preferred subject_is_issuer from a Verifier perspective? If I don't care I would not put the constraint in to begin with as it is a MAY.

I do see a use case for the inverse of required, basically stating that a subject is not the issuer.

I guess this comment is applicable to all relational constraints.

nklomp replied:
Yeah, but then you are expressing it at a per VC level anyway right? I guess it would make it a bit more expressive, but if I create a definition that simply leaves out the `preferred` for the relational constraint, I would have exactly the same constraint as it is only expressing purpose/intent.

brentzundel replied:
iirc, the original set of possible values was 'required' and 'allowed', but it was changed to 'preferred' to match the other properties.
'allowed' makes much more sense as a possible option, i.e., I need to know your address and it can come some issuer or from yourself.

kimdhamilton replied:
@brentzundel please create an issue

nklomp commented on spec/spec.md:
```diff
@@ -836,7 +877,298 @@ processing-related rules above:
+## JSON-LD Framing Feature
+
+The JSON-LD Framing [[ref:Feature]] extends [ref:Presentation Definition]] to support
+extended with the JSON-LD document framing.
```
Does it need mentioning that this depends on `limit_disclosure` being preferred or required?

bumblefudge replied:
It might help the LD-unfamiliar reader to add a sentence like: 
`LD framing objects are required for some forms of selective disclosure presentation.`
I'm not sure it is precise to describe it as a dependence in either direction, though-- you might want to include a framing object regardless of selective disclosure, just because you like RDF. who knows why people do the things they do!

nklomp replied:
Yeah agree. Although you mention `some forms of selective disclosure` already, maybe best to replace `required' with `are being`, to make it more clear that frames typically are optional even in case of selective disclosure in PEX

nklomp reviewed (commented):
Some small remarks/questions.

Unfortunately I couldn't attend last meeting, so hopefully not a stupid remark/question.
What is the exact use case for preferred relational constraints that are optional to begin with?

For selective disclosure the use case is really obvious from a Verifier Perspective. _I want to receive as little information as possible if you support selective disclosure, but I am okay with getting everything if you don't support it._

How would that work for subject_is_holder as example?. Basically you would be stating, "I don't care", which you could also express by not including it in the first place as it is optional.

kimdhamilton commented on spec/spec.md:
```diff
@@ -180,7 +185,20 @@ nested. See [Submission Requirement Rules](#submission-requirement-rules).
+one or more properties on one or more objects. 
+
+Processing entities are not required to support [[ref:Features]] on top of the 
+base [[ref:Feature]].
```
Thanks @nklomp, these are good questions, and we may need to bring these in as follow-up issues (we're trying to control the size/scope of this PR). Let's discuss tomorrow and figure out if it can be easily addressed in this PR or if it needs to be tracked in a separate issue

kimdhamilton replied:
@nklomp please create issue

brentzundel commented on spec/spec.md:
```diff
@@ -396,274 +371,339 @@ values, and an explanation why a certain item or set of data is being requested:
+  [Linked Data Cryptographic Suite Registry](https://w3c-ccg.github.io/ld-cryptosuite-registry/).
+
+
+## Submission Requirement Feature
```
would it work to nest all of the features into their own section?

kimdhamilton replied:
TODO @kimdhamilton

kimdhamilton replied:
TODO @kimdhamilton

brentzundel reviewed (commented):
overall this looks really great. I like the new structure and appreciate the work put in to make the spec better.

kimdhamilton commented on spec/spec.md:
```diff
@@ -396,274 +371,339 @@ values, and an explanation why a certain item or set of data is being requested:
+VP         | top-level
+CHAPI      | `$.data`
+
+## Claim Format Designations
```
TODO @kimdhamilton: move to right before json schemas
